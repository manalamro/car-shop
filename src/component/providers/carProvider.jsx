import React, { useState } from "react";
import {CarsArray} from '../../Data/data' // CarsArray Should be between curly braket '{}'.

export const CarsContext = React.createContext({});  
const getLoggedInUser = () => {
  const localUser = JSON.parse(localStorage.getItem('vsn-user'));
  return localUser || null;
}

const CarsProvider = (props) => {
  const [cars,setCars]= useState(CarsArray);
  const [user, setUser] = useState(getLoggedInUser());
  const handleUserLogin = (loggedInUser) => {
    setUser(loggedInUser);
    localStorage.setItem('vsn-user', JSON.stringify(loggedInUser));
  }
  //function that bring and read the last data before add the new one so that
    // its will be clear that we add new post.

  const fetchCars = () => {
    const carsFromStorage = JSON.parse(localStorage.cars || '[]');
    if (carsFromStorage.length === 0) {
      localStorage.cars = JSON.stringify(CarsArray);
      setCars(CarsArray);
    } else {
      setCars(carsFromStorage);
    }
  }

  
  return (
    <CarsContext.Provider value={{user ,cars,handleUserLogin, fetchCars, setCars}}>
      {props.children}
    </CarsContext.Provider>
  );
};

export default CarsProvider;