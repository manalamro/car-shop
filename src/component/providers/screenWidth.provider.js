import { useEffect, useState } from 'react';
// Custom hook to get the width of the screen
function useScreenWidth() {
  const [screenWidth, setScreenWidth] = useState(window.innerWidth);

  const handleResize = () => {
    setScreenWidth(window.innerWidth);
  };

  useEffect(() => {
    // Add event listener to update width on resize
    window.addEventListener('resize', handleResize);
    // Clean up event listener on unmount
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return screenWidth;
}


export default useScreenWidth;
