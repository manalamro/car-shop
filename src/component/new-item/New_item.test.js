import React from "react";
import { render, fireEvent } from "@testing-library/react";
import New_item from "./new-item.component"; // Adjust if needed
import { CarsContext } from "../providers/carProvider";

const mockContextValue = {
  user: { role: "admin" },
};

describe("New_item Component", () => {
  test("calls handleSubmit when Add button is clicked", () => {
    const mockSubmit = jest.fn();

    const { getByRole } = render(
      <CarsContext.Provider value={mockContextValue}>
        <New_item handleSubmit={mockSubmit} />
      </CarsContext.Provider>
    );

    const form = document.querySelector("form");
    fireEvent.change(form.querySelector('input[name="car_color"]'), {
      target: { value: "Red" },
    });
    fireEvent.change(form.querySelector('input[name="PayPerMonth"]'), {
      target: { value: "300" },
    });
    fireEvent.change(form.querySelector('input[name="car_meter"]'), {
      target: { value: "20000" },
    });
    fireEvent.change(form.querySelector('input[name="car_year"]'), {
      target: { value: "2023" },
    });
    fireEvent.change(form.querySelector('input[name="car_Make"]'), {
      target: { value: "Toyota" },
    });
    fireEvent.change(form.querySelector('input[name="imgURL"]'), {
      target: { value: "img.jpg" },
    });

    fireEvent.change(form.querySelector('select[name="Car_status"]'), {
      target: { value: "For Sale" },
    });
    fireEvent.change(form.querySelector('select[name="cars_brand"]'), {
      target: { value: "Golf Cabriolet" },
    });
    fireEvent.change(form.querySelector('select[name="cars_transmission"]'), {
      target: { value: "Manual" },
    });
    fireEvent.change(form.querySelector('select[name="cars_EngineFuel"]'), {
      target: { value: "Diesel" },
    });
    fireEvent.change(form.querySelector('select[name="CarType"]'), {
      target: { value: "sedan" },
    });

    fireEvent.submit(form);

    expect(mockSubmit).toHaveBeenCalledTimes(1);
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <CarsContext.Provider value={mockContextValue}>
        <New_item handleSubmit={() => {}} />
      </CarsContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
