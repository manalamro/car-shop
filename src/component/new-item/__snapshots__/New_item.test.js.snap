// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New_item Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <form
      class="form"
    >
      <div
        class="column-item"
      >
        <label>
          Car Color
        </label>
        <input
          name="car_color"
          required=""
          type="text"
        />
        <label>
          PayPerMonth
        </label>
        <input
          name="PayPerMonth"
          required=""
          type="text"
        />
        <label>
          Car Odometer
        </label>
        <input
          name="car_meter"
          required=""
          type="text"
        />
        <label>
          Car year
        </label>
        <input
          name="car_year"
          required=""
          type="number"
        />
        <label>
          Car Make
        </label>
        <input
          name="car_Make"
          required=""
          type="text"
        />
        <label>
          Car Image
        </label>
        <input
          name="imgURL"
          required=""
          type="text"
        />
        <label>
          Car status
        </label>
        <select
          name="Car_status"
          required=""
        >
          <option
            value="For Sale"
          >
            For Sale
          </option>
          <option
            value="Not For Sale"
          >
            Not For Sale
          </option>
        </select>
        <label>
          Car Brand
        </label>
        <select
          name="cars_brand"
          required=""
        >
          <option
            value="Golf Cabriolet"
          >
            Golf Cabriolet
          </option>
          <option
            value="Beetle"
          >
            Beetle
          </option>
          <option
            value="Arteon"
          >
            Arteon
          </option>
          <option
            value="Shooting Brake"
          >
            Shooting Brake
          </option>
          <option
            value="Volkswagen CC"
          >
            Volkswagen CC
          </option>
          <option
            value="BMW-i8"
          >
            BMW-i8
          </option>
          <option
            value="G-class"
          >
            G-class
          </option>
          <option
            value="Avalon Hybrid"
          >
            Avalon Hybrid
          </option>
          <option
            value="Aventador"
          >
            Aventador
          </option>
        </select>
        <label>
          Car transmission
        </label>
        <select
          name="cars_transmission"
          required=""
        >
          <option
            value="Manual"
          >
            Manual
          </option>
          <option
            value="Automatic"
          >
            Automatic
          </option>
        </select>
        <label>
          Car EngineFuel
        </label>
        <select
          name="cars_EngineFuel"
          required=""
        >
          <option
            value="Diesel"
          >
            Diesel
          </option>
          <option
            value="petrol"
          >
            petrol
          </option>
        </select>
        <label>
          Car CarType
        </label>
        <select
          name="CarType"
          required=""
        >
          <option
            value="sedan"
          >
            sedan
          </option>
          <option
            value="Universal"
          >
            Universal
          </option>
          <option
            value="CUV"
          >
            CUV
          </option>
          <option
            value="Roadster"
          >
            Roadster
          </option>
          <option
            value="sport"
          >
            sport
          </option>
        </select>
      </div>
      <div
        class="button-style"
      >
        <button
          class="submit"
          type="submit"
        >
          Add
        </button>
        <button
          class="reset"
          type="reset"
        >
          Clear
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
